config/API/custom/generic_permission_view_set.py

```diff
@@ -0,0 +1,91 @@
+from rest_framework.request import *
+from rest_framework.response import Response
+from rest_framework import status, viewsets
+from django_filters.rest_framework import DjangoFilterBackend
+
+class GenericPermissionViewSet(viewsets.ModelViewSet):
+    filter_backends = [DjangoFilterBackend]
+    filterset_fields = '__all__'
+
+    def list(self, request: Request):
+        if request.user.id is None:
+            return Response(status=status.HTTP_401_UNAUTHORIZED)
+        
+        if request.user.is_staff:
+            queryset = self.model_class.objects.all()
+        else:
+            if request.query_params.get("user") not in [None, request.user]:
+                return Response(status=status.HTTP_403_FORBIDDEN)
+            queryset = self.model_class.objects.filter(user_id = request.user.id)
+        
+        filterset = self.filterset_class(request.query_params, queryset=queryset)
+        serializer = self.serializer_class(filterset.qs, many=True)
+        return Response(serializer.data)
+    
+    def retrieve(self, request: Request, *args, **kwargs):
+        object = self.try_get_object(kwargs.get('pk'))
+        if object is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+        
+        if not request.user.is_staff and request.user.id != object.user_id:
```

Comment:
Wydaje mi się, że najpierw powinniśmy sprawdzać permisje, a potem czy obiekt istnieje, żeby nie przekazywać osobie nieautoryzowanej żadnej informacji.

Reply:
Mogę dodać na początku sprawdzenie czy request jest anonimowy i zwracać 401 we wszystkich metodach (tak jak w list() na przykład). Ale żeby sprawdzić uprawnienia to muszę najpierw pobrać obiekt z bazy, więc nie jestem w stanie najpierw sprawdzić uprawnień, a dopiero po tym czy obiekt istnieje. Mogę ewentualnie zwracać 403 zamiast 404, jeżeli obiektu nie ma, ale to jest trochę mylące.

Reply:
A masz rację w sumie, rób jak jest. Przecież jak coś nie istnieje to nie wiadomo czyje jest.